src/js/media-controller.js

```diff
@@ -38,357 +52,724 @@ const ButtonPressedKeys = ['ArrowLeft', 'ArrowRight', 'Enter', ' ', 'f', 'm', 'k
+      // Assuming seekable range already accounts for appropriate buffer room
+      const liveEdgeStartOffset = controller.hasAttribute('liveedgeoffset') 
+        // TODO: Move to no attr value
+        ? Number(controller.getAttribute('liveedgeoffset'))
+        : 10;
+      const liveEdgeStart = seekable.end(seekable.length - 1) - liveEdgeStartOffset
+      return media.currentTime >= liveEdgeStart;
+    },
+    mediaEvents: ['playing','timeupdate','progress','waiting','emptied']
+  },
+  MEDIA_IS_FULLSCREEN: {
+    get: function(controller, e) {
+      // Safari doesn't support ShadowRoot.fullscreenElement and document.fullscreenElement
+      // could be several ancestors up the tree. Use event.target instead.
+      const isSomeElementFullscreen = !!document[fullscreenApi.element];
+      const fullscreenEl = isSomeElementFullscreen && e?.target;
+      return containsComposedNode(controller.fullscreenElement, fullscreenEl);
+    },
+    // TODO: Don't miss this special event type when implementing
+    rootEvents: [fullscreenApi.event]
+  },
+  MEDIA_IS_PIP: {
+    get: function(controller, e) {
+      const media = controller.media;
+
+      if (!media) return false;
+
+      // Rely on event type for state first
+      // in case this doesn't work well for custom elements using internal <video>
+      if (e) {
+        return e.type == 'enterpictureinpicture';
+      } else {
+        const pipElement =
+          // Might need to use the root of media-container
+          // @ts-ignore
+          controller.getRootNode().pictureInPictureElement ??
+          document.pictureInPictureElement;
+        return containsComposedNode(media, pipElement);
+      }
+    },
+    mediaEvents: ['enterpictureinpicture', 'leavepictureinpicture']
+  },
+  MEDIA_IS_CASTING: {
+    // Note this relies on a customized video[is=castable-video] element.
+    get: function(controller, e) {
+      const { media } = controller;
+
+      if (!media) return false;
+
+      const castElement = globalThis.CastableVideoElement?.castElement;
+      let castState = containsComposedNode(media, castElement);
+
+      // While the cast is connecting set media-is-cast="connecting"
+      if (e?.type === 'castchange' && e?.detail === 'CONNECTING') {
+        castState = 'connecting';
+      }
+
+      return castState;
+    },
+    mediaEvents: ['entercast','leavecast','castchange']
+  },
+  MEDIA_AIRPLAY_UNAVAILABLE: {
+    // NOTE: only adding this if airplay is supported, in part to avoid unnecessary battery consumption per
+    // Apple docs recommendations (See: https://developer.apple.com/documentation/webkitjs/adding_an_airplay_button_to_your_safari_media_controls)
+    // For a more advanced solution, we could monitor for media state receivers that "care" about airplay support and add/remove
+    // whenever these are added/removed. (CJP)
+    // NOTE: I don't think only adding this if supported helps, so making this match others. 
```

Comment:
@cjpillsbury see this specific update, allowing it not be selectively added as a listener like it was before.